Tests without strings

paldepind opened the pull request:
Hello, again @polytypic. As I said in my other PR this library is really nice. It's very rare to see a library this impressive and it clear that a lot of work has been put into it 👍 

I was slightly turned off by the way the tests are written as strings. So this PR is an attempt to fix that. I believe the approach in this PR gives the benefits from using strings but without the downsides. The best of both worlds if you will 😄 

This PR has the following properties

* Tests are written in _almost_ in the exact same way as before
* Output is shown exactly as before

Tests no longer rely on strings nor `eval`. This has the following benefits

* Synax highlighting works for the test code. Similarly, any other editing enhancing feature present in ones editor (automatic indentation, autocompletion, etc) isn't turned off due to the strings.
* No need to escape with double backslashes in the `L.matches` tests that uses regexes
* No need to hackisly inject values into the test expression with `${show(value)}`
* The tests seem to run a lot faster. Probably due to the elimination of `eval`.

If you like this approach I will do the following

* Convert the remaining places where `testEq` is still called with a string
* Do the same thing for `testThrows`

paldepind commented on test/tests.js:
```diff
@@ -137,26 +121,44 @@ function toggleEnv() {
+    // const exprTy = expr.replace(/\bL\.([a-zA-Z0-9]*)/g, 'T.$1(L.$1)')
+    // const typed = run(exprTy)
+    // if (!equals(actual, typed))
+    //   throw Error(`Typed: ${show(typed)}, actual: ${show(actual)}`)
```
This will also have to be fixed. What is the purpose of the code?

polytypic replied:
The [types](https://github.com/calmm-js/partial.lenses/blob/master/test/types.js) module defines basic contracts for the library exports.  The above code wraps `L.x` operations in the test code with contract checking code `T.x(L.x)` and then runs the test.  This is somewhat difficult to do with thunks.  One possibility would be to have the types module actually export the library with the contracts in place (e.g. a bit like in [here](https://github.com/polytypic/prettier-printer/blob/master/src/prettier-printer.js) although in PL I don't want to export the contracts as they are for tests), parameterizing the unit test code with the name of the file from which to import the library and then running the tests both with and without the contracts.

paldepind commented on test/tests.js:
```diff
@@ -341,21 +343,21 @@ describe(`L.find`, () => {
-  testEq(`L.get([], [[{x: {y: 101}}]])`, [[{x: {y: 101}}]])
+  testEq(() => L.get([], [[{x: {y: 101}}]]), [[{x: {y: 101}}]])
   testEq(
     `L.get([0, L.findWith("x"), L.identity, "y", []], [[{x: {y: 101}}]])`,
```
I still need to convert some cases away from strings.

paldepind commented on test/tests.js:
```diff
@@ -137,26 +121,44 @@ function toggleEnv() {
+    console.log("ERROR: testEq still called with string.")
+    console.log(expr)
+    return
+  }
```
This will be removed once `testEq` is never called with a string.

polytypic commented:
Cool.  This has actually been on my mind for some time.  I recently started using a similar technique in some of my other library, e.g. in Partial Lenses Validation [tests](https://github.com/calmm-js/partial.lenses.validation/blob/master/test/tests.js), but haven't yet taken the time to convert the Partial Lenses tests.

paldepind commented:
@polytypic 

> This has actually been on my mind for some time.

😄 

I've just updated the PR. All instances of `testEq` has been converted to thunks and `testThrows` have been converted as well.

I've implemented a solution to the type contracts so that they are tested as well. I create a version of the library with the contracts in place an make the code use it by swapping the value `L`. It's a bit hacky but no more than the old solution.

The only remaining problem is that I now have a failing test 😒 . 

```js
L.seq
  collectM(L.seq(1, 0, 2), ['b', 'a', 'c']) => ["a","b","c"]:
Error: Expected object with keys chain,of,ap,map, got [object Object]
```

It's from the test [here](https://github.com/calmm-js/partial.lenses/pull/154/files#diff-38dcab2adf8a7e3431829d8daca050fcR1231). It seems that it is the `chain` method that is missing. I don't see how my changes have caused that. Do you have any ideas?

polytypic commented:
> It seems that it is the chain method that is missing. I don't see how my changes have caused that. Do you have any ideas?

It turns out the contract for `toFunction` was incomplete.  It is fixed in the master now, so if you rebase then tests should pass.

> I've implemented a solution to the type contracts so that they are tested as well. I create a version of the library with the contracts in place an make the code use it by swapping the value L. It's a bit hacky but no more than the old solution.

Looks ok to me.  As a bonus the contract check is now a bit more thorough as it covers all `L.*` usages that are not immediately evaluated (not just the code strings as it was before). 👍 

I noticed ESLint gave a couple of unused variable warnings.  They can be suppressed by prepending `_` to the variable names.  I've also started using Prettier in this project (although the strings weren't formatted by Prettier before).  So, it is ok to just reformat the code with Prettier (it will get rid of the other ESLint warnings).

paldepind commented:
I rebased on master. All tests pass now. I fixed the ESLint errors and applied Prettier. The PR should be mergeable now.

> the strings weren't formatted by Prettier before

Yet another benefit 😄

polytypic commented:
Looks good to me, so I'll merge.  Thank you very much for the contribution!

paldepind commented:
You're welcome 😄
